File: app/src/main/java/com/duckduckgo/mobile/android/duckduckgo/ui/editbookmark/EditBookmarkDialogFragment.java
package com.duckduckgo.mobile.android.duckduckgo.ui.editbookmark;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.View;
import android.widget.EditText;

import com.duckduckgo.mobile.android.duckduckgo.R;
import com.duckduckgo.mobile.android.duckduckgo.ui.bookmarks.BookmarkEntity;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 *    Copyright 2017 DuckDuckGo
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

public class EditBookmarkDialogFragment extends AppCompatDialogFragment {

    public static final String TAG = EditBookmarkDialogFragment.class.getSimpleName();

    public static EditBookmarkDialogFragment newInstance(int titleResId, @NonNull BookmarkEntity bookmark) {
        EditBookmarkDialogFragment dialog = new EditBookmarkDialogFragment();
        Bundle args = new Bundle();
        args.putInt(EXTRA_TITLE, titleResId);
        args.putParcelable(EXTRA_BOOKMARK, bookmark);
        dialog.setArguments(args);
        return dialog;
    }

    private static final String EXTRA_TITLE = "extra_title";
    private static final String EXTRA_BOOKMARK = "extra_bookmark";

    public interface OnEditBookmarkListener {
        void onBookmarkEdited(BookmarkEntity bookmark);
    }

    @BindView(R.id.dialog_edit_bookmark_name_edit_text)
    EditText nameEditText;

    @BindView(R.id.dialog_edit_bookmark_url_edit_text)
    EditText urlEditText;

    private OnEditBookmarkListener onEditBookmarkListener;

    private int titleResId;
    private BookmarkEntity bookmarkEntity;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            onEditBookmarkListener = (OnEditBookmarkListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString() + " must implement " + OnEditBookmarkListener.class.getSimpleName());
        }
    }

    @Override
    public void onDetach() {
        onEditBookmarkListener = null;
        super.onDetach();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        titleResId = getArguments().getInt(EXTRA_TITLE);
        bookmarkEntity = getArguments().getParcelable(EXTRA_BOOKMARK);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        View rootView = View.inflate(getContext(), R.layout.dialog_edit_bookmark, null);
        ButterKnife.bind(this, rootView);
        initUI();
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext())
                .setTitle(titleResId)
                .setView(rootView)
                .setPositiveButton(R.string.bookmark_dialog_save_button, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        onSaveButtonClick();
                    }
                })
                .setNegativeButton(R.string.bookmark_dialog_cancel_button, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dismiss();
                    }
                });
        return builder.create();
    }

    private void initUI() {
        nameEditText.setText(bookmarkEntity.getName());
        urlEditText.setText(bookmarkEntity.getUrl());
    }

    private void onSaveButtonClick() {
        onEditBookmarkListener.onBookmarkEdited(getEditedBookmark());
    }

    private BookmarkEntity getEditedBookmark() {
        bookmarkEntity.setName(nameEditText.getText().toString());
        bookmarkEntity.setUrl(urlEditText.getText().toString());
        return bookmarkEntity;
    }
}

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/EditBookmarkDialogFragment.kt
/*
 * Copyright (c) 2018 DuckDuckGo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.duckduckgo.app.bookmarks.ui

import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import com.duckduckgo.app.bookmarks.model.SavedSite
import com.duckduckgo.app.browser.R
import com.duckduckgo.app.global.view.hideKeyboard
import com.duckduckgo.app.global.view.showKeyboard

class EditBookmarkDialogFragment : DialogFragment() {

    interface EditBookmarkListener {
        fun onSavedSiteEdited(savedSite: SavedSite)
    }

    var listener: EditBookmarkListener? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val rootView = View.inflate(activity, R.layout.edit_bookmark, null)
        val titleInput = rootView.findViewById<EditText>(R.id.titleInput)
        val urlInput = rootView.findViewById<EditText>(R.id.urlInput)

        val alertBuilder = AlertDialog.Builder(requireActivity())
            .setView(rootView)
            .setTitle(R.string.bookmarkDialogTitleEdit)
            .setPositiveButton(R.string.dialogSave) { _, _ ->
                userAcceptedDialog(titleInput, urlInput)
            }

        validateBundleArguments()

        populateFields(titleInput, urlInput)

        val alert = alertBuilder.create()
        showKeyboard(titleInput, alert)
        return alert
    }

    private fun userAcceptedDialog(titleInput: EditText, urlInput: EditText) {
        when (val savedSite = getSavedSite()) {
            is SavedSite.Bookmark -> {
                listener?.onSavedSiteEdited(
                    savedSite.copy(title = titleInput.text.toString(), url = urlInput.text.toString())
                )
            }
            is SavedSite.Favorite -> {
                listener?.onSavedSiteEdited(
                    savedSite.copy(title = titleInput.text.toString(), url = urlInput.text.toString())
                )
            }
        }
        titleInput.hideKeyboard()
    }

    private fun showKeyboard(titleInput: EditText, alert: AlertDialog) {
        titleInput.setSelection(titleInput.text.length)
        titleInput.showKeyboard()
        alert.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
    }

    private fun populateFields(titleInput: EditText, urlInput: EditText) {
        titleInput.setText(getExistingTitle())
        urlInput.setText(getExistingUrl())
    }

    private fun getSavedSite(): SavedSite = requireArguments().getSerializable(KEY_SAVED_SITE) as SavedSite
    private fun getExistingId(): Long = getSavedSite().id
    private fun getExistingTitle(): String = getSavedSite().title
    private fun getExistingUrl(): String = getSavedSite().url

    private fun validateBundleArguments() {
        if (arguments == null) throw IllegalArgumentException("Missing arguments bundle")
        val args = requireArguments()
        if (!args.containsKey(KEY_SAVED_SITE)) {
            throw IllegalArgumentException("Bundle arguments required [KEY_PREEXISTING_TITLE, KEY_PREEXISTING_URL]")
        }
    }

    companion object {
        private const val KEY_SAVED_SITE = "KEY_SAVED_SITE"

        fun instance(savedSite: SavedSite): EditBookmarkDialogFragment {

            val dialog = EditBookmarkDialogFragment()
            val bundle = Bundle()

            bundle.putSerializable(KEY_SAVED_SITE, savedSite)
            dialog.arguments = bundle
            return dialog
        }
    }

}
